### PR TITLE
Implement getPendingCompactions in HttpManagementProxy

### LINKED ISSUE #1382: Implement getPendingCompactions() in the HttpManagementProxy
[Project board link](https://github.com/orgs/k8ssandra/projects/8/views/1?pane=issue&itemId=37693681)

Method signature:
`public int getPendingCompactions() throws JMException`

### LINKED ISSUE #1396: Implement the HttpMetricsProxy in Reaper
[Project board link](https://github.com/orgs/k8ssandra/projects/8/views/1?pane=issue&itemId=37820250)

When using the HttpManagementProxy, as metrics aren't exposed in the management API we need to query the metrics endpoint directly in the mgmt-api agent.
This endpoint runs in a separate jvm on a different port which has to be made configurable in Reaper since it can be modified in the mgmt-api config.
This endpoint being a Prometheus compliant one, we'll bring in the [Prometheus Java Client](https://github.com/prometheus/client_java) in order to interact with it.

The private `collectMetrics()` method will be implemented to match the requirements for the Prometheus client and the JmxStat class will be renamed to avoid implementation specific names. We can reuse it to store both metrics coming from JMX and from the mgmt-api.

## Definition of Done
- [x] the HttpMetricsProxy class is fully implemented
- [x] Metrics are pulled from the mgmt-api using the Prometheus client
- [x] A new config setting is added to allow modifying the mgmt-api port exposing the metrics

### PR DESCRIPTION
**Pull request by adejanovski**
Fixes #1382 

Since metrics are now exposed on a different port than the mgmt-api itself, this required to go through the HttpMetricsProxy which got partially implemented for that need.
It is now able to pull metrics from the metrics endpoint and parse it into JmxStat objects (which will need to be later refactored to a new name).
Only tpstats are considered for now as the others are not needed for segments orchestration.

### CODE REVIEW

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java
Comment (burmanm) [inline src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java]: issue: This will cause serious load on Reaper node (and the Cassandra node) as it will transfer all the metrics through network. This is the reason the filtering method was created, where you can append the name filtering to get the exact metric (and nothing else).

So the syntax is: ``http://localhost:9000/metrics?name=org_apache_metrics_threadetc`` and that would return only the org_apache_metrics_threadetc metric (exact name is required).

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java
Comment (burmanm) [inline src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java]: question: Why are these metrics converted to JmxStat instead of creating directly the GenericMetric class, which is what the methods will return? I feel like now we have to create a refactoring ticket to remove JmxStat from HttpMetrics as this class isn't doing any JMX work.

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java
Comment (burmanm) [inline src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java]: issue: This isn't calling ``collectMetrics`` with correct parameters. Should be ``collectMetrics(Optional.of(THREAD_POOL_METRICS_PREFIX))``.

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java]: Right, I was under the impression initially that the interface was exposing JmxStats as return type, but collectMetrics isn't defined in it and is a private method.
I'll get rid of JmxStat there.

File: src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java
Comment (Miles-Garnsey) [inline src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java:44]: Suggestion: Not sure about this. I'd almost rather this switch (between HTTP and JMX metrics proxies) happened in ReaperApplication. I don't think this interface should know whether it is working on a JMX or HTTP backend.

File: src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java
Comment (Miles-Garnsey) [inline src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java:67]: Suggestion: Would it be worth making this non-static so you don't have to pass `node`?

File: src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java
Comment (Miles-Garnsey) [inline src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java:90]: Issue: would be good to see a test for this. Copilot should be able to write something reasonable for it quite quickly I think.

Same for the other methods here.

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java
Comment (Miles-Garnsey) [inline src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java]: Issue: I assume this only gets hit if the tpStatsMetrics doesn't have a `CompactionExecutor` item with scope = `PendingTasks`. I think in that case maybe this is an error and we should log a warning and throw, what do you think?

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java
Comment (Miles-Garnsey) [inline src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java:113]: Suggestion: OK, this feels a bit weird. Our MetricsProxy takes a `HttpCassandraManagementProxy` as a argument to the constructor, but `HttpCassandraManagementProxy` also has an instance of the `MetricsProxy`, it feels a bit tangled. Is it possible to get some clearer separation between these classes?

Possibly this is an item for later refactoring, but I had to ask...

Comment (Miles-Garnsey) [reply src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java:113]: It actually looks like the proxy is called only for this line of code:

```
String url = String.format("http://%s:%s/metrics", proxy.getHost(), proxy.getMetricsPort());
```

It seems to me like maybe the HttpCassandraManagementProxy shouldn't know about the metrics port at all. And the MetricsProxy should maybe take the host as an argument to it's constructor instead of taking the whole ICassandraManagementProxy.

I haven't looked at what this would imply for the JMX side of things, but think we should try to do something a bit cleaner here.

**Review by Miles-Garnsey — CHANGES_REQUESTED** [review]
I've done a preliminary review, I'll look more tomorrow. I have a few requested changes, but most aren't really blocking. 

Given the tests are passing, I'm almost happy to approve immediately, but given the size of this PR I want to take another look over it. I can see why it took a little while!

File: src/server/src/test/java/io/cassandrareaper/acceptance/BasicSteps.java
Comment (Miles-Garnsey) [inline src/server/src/test/java/io/cassandrareaper/acceptance/BasicSteps.java:2076]: Issue: do we want the test to fail out at this point? I think maybe we do.

Comment (Miles-Garnsey) [reply src/server/src/test/java/io/cassandrareaper/acceptance/BasicSteps.java:2076]: OR, maybe this isn't an error and should be logged as a warning/debug, since in a real environment I guess sometimes messages won't have been dropped...

File: src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java:44]: A lot of the changes here support the fact that the management proxy now invokes the metrics proxy. It wasn't the case before as the metrics proxy was using the management proxy to get the metrics (through JMX). It's now the other way around with the HTTP flavor as we're not getting the metrics through the management api client, but through a separate endpoint.
It's a fundamental change between the JMX and HTTP implementations which I couldn't do differently without a deep refactor.

File: src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java:67]: That doesn't seem to be possible since the creation of a MetricsProxy instance is done through a static method, which can't initialize a member of the implementation class.

File: src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java:90]: sounds good 👍

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java:113]: I tried to untangle them but failed so far, it's a fairly involved change.
I'd recommend to create a refactoring ticket to work this out.
And I agree that we should not keep it like this over longer term. JMX and HTTP differ pretty heavily here due to the separate endpoints for operations and metrics in the management api. It requires some time and thought to review the concept of MetricsProxy here.

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java]: Fair enough, I thought I was reproducing the behavior of the JMX implementation, but it turns out we re-throw JMExceptions as ReaperExceptions, which should include the case where we can't find the metric we're looking for. I'll do the same here and throw a ReaperException if the metric can't be found.

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java
Comment (Miles-Garnsey) [reply src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java:113]: OK, can you please create a refactoring ticket before merging.

File: src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java
Comment (Miles-Garnsey) [reply src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java:44]: We can roll this into the refactor that we need to do to disentangle these classes down the track. Please include a note in your refactoring ticket that the `create()` method probably shouldn't be the thing deciding whether we want a HTTP or JMX backed implementation.

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java
Comment (Miles-Garnsey) [reply src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java]: I don't know if it is just the diff issues I'm seeing, but this doesn't appear to have been updated to throw an exception?

**Review by Miles-Garnsey — CHANGES_REQUESTED** [review]
I mentioned on slack, but I'm having issues with reviewing this due to noise in the diff (I think the last rebase I did was responsible, need to discuss how to do this better next time).

Can we maybe rebase this against the current integration branch so I have clearer visibility?

File: src/server/src/test/java/io/cassandrareaper/acceptance/BasicSteps.java
Comment (adejanovski) [reply src/server/src/test/java/io/cassandrareaper/acceptance/BasicSteps.java:2076]: I'm failing fast now

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java
Comment (Miles-Garnsey) [inline src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java:100]: Issue: can we choose between throwing an exception and returning an empty and do one or the other consistently in these methods. 

Suggestion: it might be good to put a TODO on these (as that gets picked up by IDEs more, and is also something people can search for), and it would also be good to raise a ticket to complete these methods too.

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java
Comment (Miles-Garnsey) [inline src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java:110]: Suggestion: It would be good to see a test here, but this is non blocking as technically I  think this method will be tested via other tests (e.g.`testCollectTpPendingTasks()`).

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java
Comment (Miles-Garnsey) [inline src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java:122]: Suggestion: non blocking but I don't love the use of strings where a more specific type is available. In this case `HttpUrl` would give you some nicer, more expressive methods.

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java
Comment (Miles-Garnsey) [inline src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java:145]: Question: If there is no MetricType with value `ThreadPools` is this an error that should be thrown, or is it an expected condition which should return an empty list?

Comment (Miles-Garnsey) [reply src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java:145]: Also, if the answer is that this should throw, can you go back through each method to ensure that they all have consistent behaviour in that regard.

File: src/server/src/test/resources/cassandra-reaper-http-at.yaml
Comment (Miles-Garnsey) [inline src/server/src/test/resources/cassandra-reaper-http-at.yaml]: Question: Is all of this new stuff required?

**Review by Miles-Garnsey — CHANGES_REQUESTED** [review]
Some more feedback now that I can see what's happening here a bit better. I might have yet more tomorrow once I've gone through this in more depth again. 

Sorry, this is a relatively large PR, and the amount of refactoring required (because we're making compromises to get it across the line) makes the logic much harder to make out.

File: src/server/src/test/resources/cassandra-reaper-http-at.yaml
Comment (adejanovski) [reply src/server/src/test/resources/cassandra-reaper-http-at.yaml]: yep, we need the driver configuration to connect to Cassandra.

File: src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java:44]: https://github.com/thelastpickle/cassandra-reaper/issues/1423

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java:113]: https://github.com/thelastpickle/cassandra-reaper/issues/1423

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java:100]: sure thing, I'll return an empty list here.
Also I think it's the only method that needs to be implemented still, the others are done (I've removed the misleading comments there).

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java:110]: yeah, it's tested through the other methods. Testing it again would be slightly redundant.

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java:122]: ok, doing that.

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java:145]: I'm not sure about the runtime behavior here, so I'd go with returning an empty list if no metric matches the requested metric type.

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java]: It should appear after the branch was squashed and rebased.

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java
Comment (emerkle826) [reply src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java:110]: That is the case. All of the public `collectXYZ` methods just call this method, so testing them tests this method.

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java
Comment (Miles-Garnsey) [reply src/server/src/main/java/io/cassandrareaper/management/http/HttpMetricsProxy.java]: I think this has been resolved by allowing the method to take an optional of the metric name and filter on that.

File: src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java
Comment (Miles-Garnsey) [reply src/server/src/main/java/io/cassandrareaper/management/MetricsProxy.java:90]: Appears to have been added

File: src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java
Comment (Miles-Garnsey) [reply src/server/src/main/java/io/cassandrareaper/management/http/HttpCassandraManagementProxy.java:113]: Alex has said he's creating a refactoring ticket.

File: src/server/src/test/resources/cassandra-reaper-http-at.yaml
Comment (Miles-Garnsey) [reply src/server/src/test/resources/cassandra-reaper-http-at.yaml]: I'm still waiting on a response here.

**Review by Miles-Garnsey — APPROVED** [review]
@adejanovski is going to fix a duplicated block in the Reaper config for the acceptance tests, and also link to the metrics proxy refactoring ticket he raised.

Otherwise, this is approved.

**Comment by adejanovski** [comment]
Here's the follow up ticket: #1423

**Pull request merged**
